Firestore/core/src/model/value_util.cc

```diff
@@ -736,6 +753,62 @@ bool IsMaxValue(const google_firestore_v1_Value& value) {
              kRawMaxValueFieldValue;
 }
 
+int64_t IndexOfKey(const google_firestore_v1_MapValue& mapValue,
+                   const char* kRawTypeValueFieldKey,
+                   pb_bytes_array_s* kTypeValueFieldKey) {
+  for (pb_size_t i = 0; i < mapValue.fields_count; i++) {
+    if (mapValue.fields[i].key == kTypeValueFieldKey ||
+        nanopb::MakeStringView(mapValue.fields[i].key) ==
+            kRawTypeValueFieldKey) {
+      return i;
+    }
+  }
+
+  return -1;
+}
+
+bool IsVectorValue(const google_firestore_v1_Value& value) {
+  if (value.which_value_type != google_firestore_v1_Value_map_value_tag) {
+    return false;
+  }
+
+  if (value.map_value.fields_count < 2) {
+    return false;
+  }
+
+  int64_t typeFieldIndex = -1;
+  if ((typeFieldIndex = IndexOfKey(value.map_value, kRawTypeValueFieldKey,
+                                   kTypeValueFieldKey)) < 0) {
+    return false;
+  }
+
+  if (value.map_value.fields[typeFieldIndex].value.which_value_type !=
+      google_firestore_v1_Value_string_value_tag) {
+    return false;
+  }
+
+  // Comparing the pointer address, then actual content if addresses are
+  // different.
+  return value.map_value.fields[typeFieldIndex].value.string_value ==
+             kVectorTypeFieldValue ||
+         nanopb::MakeStringView(
+             value.map_value.fields[typeFieldIndex].value.string_value) ==
+             kRawVectorTypeFieldValue;
+
+  int64_t valueFieldIndex = -1;
```

Comment:
This block is unreachable, isn't it?

Reply:
Good catch. There were several bugs in this function.